understand-tidb/introduction-of-tidb-architecture.md

```diff
@@ -1,10 +1,157 @@
 
-Agenda:
+Understanding TiDB talks about what is the architecture of TiDB, what modules it consists of, and what is the responsibility of each module.
 
-* TiDB architecture: tidb-server, pd-server, tikv-server
-* tidb-server internal architecture
-* tidb code structure
+## TiDB Architecture
 
+When people refer to TiDB, they usually refer to the entire TiDB distributed database that includes three components: the TiDB stateless server, the Placement Driver (PD) server, and the storage server, TiKV or TiFlash. The TiDB server does not store data; it only computes and processes SQL queries. The PD server is the managing component of the entire cluster. The storage server is responsible for persistently storing data.
 
+Let's see an architecture graph from TiDB stateless server's perspective.
 
+![tidb-architecture](https://user-images.githubusercontent.com/18818196/121637571-28acae80-cabc-11eb-970e-14af8a7a6246.png)
+
+As you can see, TiDB is a SQL engine that supports the MySQL protocol with some kind of distributed KV storage engine that supports transactions as the underlying storage.
+
+Here come three significant questions.
+
+1. How to support MySQL protocol?
+2. How to communicate with storage engine, store and load data?
+3. How to implement SQL functions?
+
+This section will start with a few of brief descriptions of what modules TiDB has and what they do, and then put them together with these three questions.
+
+## Code Structure
+
+TiDB source code is fully hosted on Github, you can see all the information from the [repository homepage](https://github.com/pingcap/tidb). The whole repository is developed in Golang and divided into many Packages according to functional modules. Through dependency analysis tools, you can see the dependencies between packages inside the project.
+
+Most of the packages export services in the form of interfaces, and most of the functionality is concentrated in one package. But there are packages that provide basic functionality and are dependent on many packages, so these packages need special attention.
+
+The main file of TiDB is in `tidb-server/main.go`, which defines how the service is started. The Build method for the entire project can be found in the `Makefile`.
+
+In addition to the code, there are many test cases, which can be found in `xx_test.go`. There is also toolkit under the `cmd` directory for doing performance tests or constructing test data.
+
+### Module Structure
+
+TiDB has a number of modules. Table below is an overview that shows what each module does, and if you want to see the code for the relevant function, you can find the corresponding module directly.
+
+| Package                | Introduction                                                                                                                                                                                                                                                                                   |
```

Comment:
There is a big change to the packages, especially for the `cmd` and `util` directories. `util` has nearly 50 packages, which is almost impossible to maintain. Maybe we don't need to list all the packages, or just give an outline of `cmd` and `util`?

Reply:
Make sense. I will remove some of sub packages in this run. But maybe we'd better structure our codebase to be clear to express.